src/game/memory.h

```diff
@@ -48,16 +36,90 @@ void *segmented_to_virtual(const void *addr);
 void *virtual_to_segmented(u32 segment, const void *addr);
 void move_segment_table_to_dmem(void);
 
-void main_pool_init(void *start, void *end);
-void *main_pool_alloc(u32 size, u32 side);
-u32 main_pool_free(void *addr);
-void *main_pool_realloc(void *addr, u32 size);
+struct MainPoolRegion {
+    u8* start;
+    u8* end;
+};
+
+#define MAIN_POOL_REGIONS_COUNT 1
+
+struct MainPoolContext {
+    struct MainPoolRegion regions[MAIN_POOL_REGIONS_COUNT];
+};
+
+extern struct MainPoolContext gMainPool;
+
+// takes the first 'size' bytes from 'region'
+static inline void* main_pool_region_try_alloc_from_start(struct MainPoolRegion* region, u32 size) {
+    u8* buf = region->start;
+    u8* newStart = buf + size;
+    if (!buf)
+        __builtin_unreachable();
+
+    region->start = newStart;
+    return buf;
+}
+
+/*
+ Main Pool is a trivial allocator that is managing multiple 'regions' of memory.
+ 'Region' is a contiguous block of memory available for main pool use.
+ It behaves similarly to an array of AllocOnly pools from vanilla SM64 by
+ "cutting" the start of the "region" when an allocation is made and returning the
+ pointer to the start of the initial "region".
+
+ Here is a simple visual example of how the memory is laid out in the main pool:
+
+ Main pool initial state is a multiple regions of memory:
+ |-------|   |----|  |-------------------|
+ If alloc(sizeof(+++++)), first region is used and the state becomes:
+ |+++++--|   |----|  |-------------------|
+  ^
+  returned pointer, no extra memory overhead
+ If afterwards alloc(sizeof(+++)) is used, it does not fit in region 1, so region 2 is used:
+ |+++++--|   |+++-|  |-------------------|
+              ^
+  returned pointer
+ */
```

Comment:
Where is this logic being applied, is it just for right side? The left side allocations don't appear to be doing this (so move this comment).

Reply:
Removed incorrect comment.